### PR TITLE
Pass FrameworkReference/PackageDownload items to NuGet

### PR DESCRIPTION
**Pull request by davkean**
This includes https://github.com/dotnet/project-system/pull/4756, so only review https://github.com/dotnet/project-system/commit/9735e850cacd353b575cfb8515258a7d7c608ec0.

Pass FrameworkReference/PackageDownload to NuGet

Fixes: #4567.

- [x] End to End verification
- [x] Tests
- [x] Need make a decision on duplicate PackageDownload - the spec calls it out but we cannot support it because CPS can't represent it.

~This is blocked on an insertion from NuGet that adds a _CollectionFrameworkReferences_ target.~ This works on [28901.261.d16.1stg]

### CODE REVIEW

**Comment by davkean** [comment]
tag @nkolev92 @nguerrera @dsplaisted

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets [hunk 1]
```diff
@@ -294,6 +294,16 @@
       <Context>ProjectSubscriptionService;BrowseObject</Context>
     </PropertyPageSchema>
 
+    <!-- Framework references -->
+    <PropertyPageSchema Include="$(ManagedXamlNeutralResourcesDirectory)FrameworkReferenceProjectBuild.xaml">
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets]: Note - these are very deliberately named "ProjectBuild" to indicate that they come from build, as opposed to evaluation, similar to the "ResolvedXXX" concept. Open to a better name, maybe "CollectedFrameworkReference" to match the "CollectXXX" targets that are called to return them?

I want a naming convention as I was extremely surprised that `PackageReference` appears to be an evaluation item but actually causes a design-time build.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/FrameworkReferenceProjectBuild.xaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  <Rule.DataSource>
+    <DataSource HasConfigurationCondition="False"
+                ItemType="FrameworkReference"
+                MSBuildTarget="CollectFrameworkReferences"
+                Persistence="ProjectFile"
+                SourceOfDefaultValue="AfterContext"
+                SourceType="TargetResults" />
+  </Rule.DataSource>
+
+  <StringProperty Name="PrivateAssets" />
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/FrameworkReferenceProjectBuild.xaml]: @nkolev92 I've already added PrivateAssets and will automatically pass it to NuGet when requested.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/RestoreComparer.TargetFrameworkInfoEqualityComparer.cs [hunk 3]
```diff
@@ -24,6 +24,18 @@ public override bool Equals(IVsTargetFrameworkInfo x, IVsTargetFrameworkInfo y)
...
+                IEnumerable<IVsReferenceItem> yFrameworkReferences = y.FrameworkReferences.Cast<IVsReferenceItem>();
+
+                if (!xFrameworkReferences.SequenceEqual(yFrameworkReferences, ReferenceItems))
+                    return false;
+
+                IEnumerable<IVsReferenceItem> xPackageDownloads = x.PackageDownloads.Cast<IVsReferenceItem>();
+                IEnumerable<IVsReferenceItem> yPackageDownloads = y.PackageDownloads.Cast<IVsReferenceItem>();
+
+                if (!xPackageDownloads.SequenceEqual(yPackageDownloads, ReferenceItems))
+                    return false;
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/RestoreComparer.TargetFrameworkInfoEqualityComparer.cs]: Does ordering of `FrameworkReferences` and `PackageDownloads` matter for the purposes of comparison?

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/RestoreComparer.TargetFrameworkInfoEqualityComparer.cs]: It doesn't, but I'd already used SequenceEqual here for the others as well. I do a follow up doing an unordered comparison for all of them.

**Comment by davkean** [comment]
This is ready for review.

File: setup/Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles/CommonFiles.swr [hunk 4]
```diff
@@ -36,13 +36,15 @@ folder "InstallDir:MSBuild\Microsoft\VisualStudio\Managed"
...
   file source="$(VisualStudioXamlRulesDir)EmbeddedResource.xaml"
+  file source="$(VisualStudioXamlRulesDir)CollectedFrameworkReference.xaml"
   file source="$(VisualStudioXamlRulesDir)Folder.xaml"
   file source="$(VisualStudioXamlRulesDir)GeneralBrowseObject.xaml"
   file source="$(VisualStudioXamlRulesDir)GeneralConfiguredBrowseObject.xaml"
   file source="$(VisualStudioXamlRulesDir)LanguageService.xaml"
   file source="$(VisualStudioXamlRulesDir)None.xaml"
   file source="$(VisualStudioXamlRulesDir)None.BrowseObject.xaml"
   file source="$(VisualStudioXamlRulesDir)NuGetRestore.xaml"
+  file source="$(VisualStudioXamlRulesDir)CollectedPackageDownload.xaml"
```
Comment (davidwengier) [inline setup/Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles/CommonFiles.swr]: Nit: With the new prefix these are no longer sorted correctly

Comment (davkean) [reply setup/Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles/CommonFiles.swr]: Yeah I noticed this when I renamed the others in a pending PR; I'll auto-gen this file or write something that will sort it in that PR.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets]: I renamed these to "CollectedXXX"

**Pull request merged**